gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ApplicationServiceImpl.java

```diff
@@ -809,60 +815,70 @@ private Single<Application> validateRedirectUris(Application application) {
-                        if (AM_V2_VERSION.equals(oAuthSettings.getSoftwareVersion())) {
-                            oAuthSettings.setSoftwareVersion(null);
-                        } else {
-                            return Single.error(new InvalidRedirectUriException());
-                        }
+                    Single<Application> error = checkRedirectUri(oAuthSettings);
+                    if (error != null) {
+                        return error;
                     }
-
-                    //check redirect_uri content
                     if (oAuthSettings.getRedirectUris() != null) {
                         for (String redirectUri : oAuthSettings.getRedirectUris()) {
-                            try {
-                                URI uri = redirectUri.contains("*") ? new URI(redirectUri) : UriBuilder.fromURIString(redirectUri).build();
-
-                                if (uri.getScheme() == null) {
-                                    return Single.error(new InvalidRedirectUriException("redirect_uri : " + redirectUri + " is malformed"));
-                                }
-
-                                final String host = isHttp(uri.getScheme()) ? uri.toURL().getHost() : uri.getHost();
-
-                                //check localhost allowed
-                                if (!domain.isRedirectUriLocalhostAllowed() && isHttp(uri.getScheme()) && UriBuilder.isLocalhost(host)) {
-                                    return Single.error(new InvalidRedirectUriException("localhost is forbidden"));
-                                }
-                                //check http scheme
-                                if (!domain.isRedirectUriUnsecuredHttpSchemeAllowed() && uri.getScheme().equalsIgnoreCase("http")) {
-                                    return Single.error(new InvalidRedirectUriException("Unsecured http scheme is forbidden"));
-                                }
-                                //check wildcard
-                                if (!domain.isRedirectUriWildcardAllowed() &&
-                                        (nonNull(uri.getPath()) && uri.getPath().contains("*") || nonNull(host) && host.contains("*"))) {
-                                    return Single.error(new InvalidRedirectUriException("Wildcard are forbidden"));
-                                }
-                                // check fragment
-                                if (uri.getFragment() != null) {
-                                    return Single.error(new InvalidRedirectUriException("redirect_uri with fragment is forbidden"));
-                                }
-                            } catch (IllegalArgumentException | URISyntaxException ex) {
-                                return Single.error(new InvalidRedirectUriException("redirect_uri : " + redirectUri + " is malformed"));
+                            Single<Application> redirectUri1 = checkRedirectContent(domain, redirectUri);
```

Comment:
suggestion: instead of providing a Singel<Application> the two methods `checkRedirectUri` and `checkRedirectContent`may throw an exception in case of error to the refactor could be at the end

```
try {
checkRedirectUri
checkRedirectConent
return SIngle.just(app)
} catch (e) {
 return Single.error(e);
}
```

Reply:
fixed